Assets/_BForBoss/_Character/Scripts/PlayerWallRunBehaviour.cs

```diff
@@ -81,50 +114,40 @@ public class PlayerWallRunBehaviour : MonoBehaviour
         #region PROPERTIES
         private Transform ChildTransform => _fpsCharacter != null ? _fpsCharacter.rootPivot : transform;
 
-        public bool IsWallRunning => _isWallRunning;
+        public bool IsWallRunning { get; private set; }
+
         #endregion
 
         #region PUBLIC_METHODS
-        public void Initialize(ECM2.Characters.Character baseCharacter, Func<Vector2> getMovementInput, Action<int> OnWallRunFinished)
+        public void Initialize(ECM2.Characters.Character baseCharacter, Func<Vector2> getMovementInput, Action<int> onWallRunFinished)
         {
             _baseCharacter = baseCharacter;
             _fpsCharacter = baseCharacter as FirstPersonPlayer;
             _movementInput = getMovementInput;
-            _OnWallRunFinished = OnWallRunFinished;
-            _mask = LayerMask.GetMask("ParkourWall");
+            _OnWallRunFinished = onWallRunFinished;
+            _parkourWallMask = LayerMask.GetMask(PARKOUR_WALL_LAYER);
         }
 
         public void Falling(Vector3 _)
         {
             if (!CanWallRun()) 
                 return;
-            RaycastHit[] hits = new RaycastHit[_directions.Length];
-            for (int i = 0; i < _directions.Length; i++)
-            {
-                Vector3 playerFacingDirection = ChildTransform.TransformDirection(_directions[i]);
-                Physics.Raycast(ChildTransform.position, playerFacingDirection, out hits[i], _wallMaxDistance, _mask);
-                if (hits[i].collider != null)
-                {
-                    Debug.DrawRay(ChildTransform.position, playerFacingDirection * hits[i].distance, Color.green);
-                }
-                else
-                {
-                    Debug.DrawRay(ChildTransform.position, playerFacingDirection * _wallMaxDistance, Color.red);
-                }
-            }
 
-            if (GetSmallestRaycastHitIfValid(hits, out RaycastHit hit) && IsClearOfGround())
+            var shouldStartWallRun = ProcessRaycasts(_startWallRunDirections, out var hit, ChildTransform, _parkourWallMask, _wallMaxDistance)
+                                     && IsClearOfGround()
+                                     && hit.collider != _lastWall;
```

Comment:
Nitpick : Probably could have this long check in its own method `private bool ShouldStartWallRun()`